src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -175,11 +185,34 @@ public Task StopAsync()
             catch (Exception ex)
             {
                 Log.UnexpectedExceptionInStop(Logger, ConnectionId, ex);
-            }
-            
+            }
+
             return Task.CompletedTask;
         }
 
+        public async Task CloseAsync()
+        {
+            if (!_closeConnectionLock.Wait(0))
+            {
+                // only one close async process can exist at a time.
+                return;
+            }
+
+            try
+            {
+                await WriteAsync(new CloseConnectionMessage(Constants.Fin));
+                await _incomingTask;
+            }
+            catch (Exception ex)
+            {
+                Log.UnexpectedExceptionInClose(Logger, ConnectionId, ex);
+            }
+
+            // wait until outcoming pipe has been drained.
+            _shutdown = true;
+            await _outgoingTask;
```

Comment:
Here is a good example of how to drain `sending` and `receiving`: https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs#L157